File: set1/challenge1.py
# http://cryptopals.com/sets/1/challenges/1/
from itertools import chain, imap, izip_longest

def pad8(bits):
    padding = 8 - len(bits)
    return padding * [0] + bits


def int2bits(integer):
    result = []
    div, mod = divmod(integer, 2)
    result.append(mod)
    while div:
        div, mod = divmod(div, 2)
        result.append(mod)
    return pad8(list(reversed(result)))


def bits2int(bits):
    size = len(bits) - 1
    return sum([b * 2 ** (size - i) for i, b in enumerate(bits)])


def hex2int(chars):
    return int(''.join(chars), 16)


def int2base64(integer):
    if integer < 26:
        return chr(ord('A') + integer)
    elif integer < 52:
        return chr(ord('a') + integer - 26)
    elif integer < 62:
        return chr(ord('0') + integer - 52)
    elif integer == 62:
        return '+'
    else:
        return '/'


def hex2base64(input):
    args = [iter(input)] * 2
    hexdigits = izip_longest(fillvalue='', *args)
    integers = imap(hex2int, hexdigits)
    bytes_iter = imap(int2bits, integers)
    bitstream = chain(*bytes_iter)
    args = [iter(bitstream)] * 6
    sixchunks = izip_longest(fillvalue=0, *args)
    integers2 = imap(bits2int, sixchunks)
    base64 = imap(int2base64, integers2)
    result = ''.join(base64)
    return result


if __name__ == '__main__':
    input = '49276d206b696c6c696e6720796f757220627261696e206c696b65206120706f69736f6e6f7573206d757368726f6f6d'
    assert hex2base64(input) == 'SSdtIGtpbGxpbmcgeW91ciBicmFpbiBsaWtlIGEgcG9pc29ub3VzIG11c2hyb29t'

File: set1/challenge1.hs
-- http://cryptopals.com/sets/1/challenges/1/
import Data.Char (chr, ord)
import Numeric (readHex)

groupBy :: Int -> [a] -> [[a]]
groupBy n [] = []
groupBy n xs = [take n xs] ++ groupBy n (drop n xs)

hex2int :: [Char] -> Int
hex2int x = fst (head (readHex x))

mods :: Int -> [Int]
mods 0 = [0]
mods i =
    let (div, mod) = divMod i 2
    in mod:mods div

pad8 :: [Int] -> [Int]
pad8 xs = if length(xs) < 8 then pad8 (0:xs) else xs

int2bits :: Int -> [Int]
int2bits i = pad8 (reverse (mods i))

bits2int :: [Int] -> Int
bits2int xs =
    let size = (length xs) - 1
    in sum [snd(pair) * (2 ^ (size - fst(pair))) | pair <- zip [0..] xs]

int2base64 :: Int -> Char
int2base64 i
    | i < 26 = chr ((ord 'A') + i)
    | i < 52 = chr ((ord 'a') + i - 26)
    | i < 62 = chr ((ord '0') + i - 52)
    | i == 62 = '+'
    | otherwise = '/'

hex2base64 :: [Char] -> [Char]
hex2base64 s =
    let sixbitsblocks = groupBy 6 (concat [int2bits (hex2int i) | i <- groupBy 2 s])
    in [int2base64 (bits2int hextet) | hextet <- sixbitsblocks]

input = "49276d206b696c6c696e6720796f757220627261696e206c696b65206120706f69736f6e6f7573206d757368726f6f6d"
expectedOutput = "SSdtIGtpbGxpbmcgeW91ciBicmFpbiBsaWtlIGEgcG9pc29ub3VzIG11c2hyb29t"

main =
    let output = (hex2base64 input)
    in if output == expectedOutput then putStrLn "Correct" else putStrLn "Incorrect"
